include/pybind11/pytypes.h

```diff
@@ -478,6 +478,42 @@ inline handle get_function(handle value) {
     return value;
 }
 
+// Reimplementation of python's dict helper functions
+
+// copied from cpython _PyDict_GetItemStringWithError
+inline PyObject * dict_getitemstring(PyObject *v, const char *key)
+{
+#if PY_MAJOR_VERSION >= 3
+    PyObject *kv, *rv;
+    kv = PyUnicode_FromString(key);
+    if (kv == NULL) {
+        throw error_already_set();
+    }
+
+    rv = PyDict_GetItemWithError(v, kv);
```

Comment:
At this point, can't you just call `dict_getitemstring`?

Reply:
Nope, still has potential to raise an exception. See https://bugs.python.org/issue35459

As copied from cpython @ https://github.com/python/cpython/blob/d5fc99873769f0d0d5c5d5d99059177a75a4e46e/Objects/dictobject.c#L1545

Reply:
My bad, I was thinking about calling your `dict_getitem`. From this line on, `dict_getitemstring` and `dict_getitem` are the same.

Reply:
Not quite the same, Py_DECREF needs to be called on the key before the error is thrown.